### PR TITLE
Fix exception on composite keys #834

### LINKED ISSUE #834: Does not work: Admin for entity with a composite key in 3.6.0
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.35.2 3.35.2 The missing Symfony Admin Generator
sonata-project/block-bundle              3.12.1 3.12.1 Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.0.1  Cache library
sonata-project/classification-bundle     3.6.1  3.6.1  Symfony SonataClassificationBundle
sonata-project/core-bundle               3.9.1  3.9.1  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.3.1  2.3.1  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  1.0.2  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.6.0  3.6.0  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.5.0  2.5.0  Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.9.0  1.9.0  Lightweight Exporter library
sonata-project/formatter-bundle          3.4.1  3.4.1  Symfony SonataFormatterBundle
sonata-project/google-authenticator      2.1.0  2.1.0  Library to integrate Google Authenticator into a PHP project
sonata-project/intl-bundle               2.4.1  2.4.1  Symfony SonataIntlBundle
sonata-project/media-bundle              3.13.0 3.13.0 Symfony SonataMediaBundle
sonata-project/user-bundle               3.6.0  4.2.1  Symfony SonataUserBundle
```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/assetic-bundle     v2.8.2  v2.8.2  Integrates Assetic into Symfony2
symfony/monolog-bundle     v2.12.1 v3.2.0  Symfony MonologBundle
symfony/phpunit-bridge     v2.8.40 v4.0.10 Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.8.0  v1.8.0  Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-ctype     v1.8.0  v1.8.0  Symfony polyfill for ctype functions
symfony/polyfill-intl-icu  v1.8.0  v1.8.0  Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.8.0  v1.8.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php54     v1.8.0  v1.8.0  Symfony polyfill backporting some PHP 5.4+ features to lower PHP versions
symfony/polyfill-php55     v1.8.0  v1.8.0  Symfony polyfill backporting some PHP 5.5+ features to lower PHP versions
symfony/polyfill-php56     v1.8.0  v1.8.0  Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.8.0  v1.8.0  Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.8.0  v1.8.0  Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.1  v3.0.1  Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.6.7  v3.2.2  Symfony SwiftmailerBundle
symfony/symfony            v2.8.40 v4.0.10 The Symfony PHP framework
```

#### PHP version

```
$ php -v
PHP 7.1.13-1+ubuntu16.04.1+deb.sury.org+1 (cli) (built: Jan  5 2018 13:26:45) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.1.13-1+ubuntu16.04.1+deb.sury.org+1, Copyright (c) 1999-2017, by Zend Technologies
    with Xdebug v2.5.5, Copyright (c) 2002-2017, by Derick Rethans

```

## Subject

I have entity with composite id: 
```
AppBundle\Entity\DomainOffice:
    type: entity
    table: app__domain_office
    id:
        Domain:                                     { associationKey: true }
        Office:                                     { associationKey: true }
    fields:
        position:                                   { type: bigint, nullable: false, options: { unsigned: false } }
    manyToOne:
        Domain:
            targetEntity:                           AppBundle\Entity\Domain
            inversedBy:                             DomainOffices
            joinColumn:                             { name: region_id, referencedColumnName: departure_city_id }
        Office:
            targetEntity:                           AppBundle\Entity\Office
            joinColumn:                             { name: office_id, referencedColumnName: id }
    lifecycleCallbacks: {  }
```
Config for admin: services.yml
```
services:
    app.admin.domain_office:
        class: AppBundle\Admin\DomainOfficeAdmin
        arguments: [~, AppBundle\Entity\DomainOffice, AppBundle:DomainSortAdmin]
        tags:
            - { name: sonata.admin, manager_type: orm, group: "Sorting", label: "Domain offices" }
        public: true
```

Admin class
```
<?php

namespace AppBundle\Admin;

use Sonata\AdminBundle\Admin\AbstractAdmin;
use Sonata\AdminBundle\Datagrid\ListMapper;
use Sonata\AdminBundle\Route\RouteCollection;

class DomainOfficeAdmin extends AbstractAdmin
{
    protected function configureRoutes(RouteCollection $collection)
    {
        $collection->remove('create');
        $collection->remove('delete');
        $collection->add('up', $this->getRouterIdParameter() . '/up');
        $collection->add('down', $this->getRouterIdParameter() . '/down');
    }

    protected function configureListFields(ListMapper $listMapper)
    {
        $listMapper
            ->add('Domain', null, ['label' => 'Domain'])
            ->add('Office', null, ['label' => 'Office'])
            ->add('position', null, ['label' => 'Position'])
            ->add('_action', null, array(
                'actions' => array(
                    'up' => array('template' => 'AppBundle:Admin:Sort/list__action_up.html.twig',),
                    'down' => array('template' => 'AppBundle:Admin:Sort/list__action_down.html.twig',),
                ),
            ));
    }
}
```

## Steps to reproduce
After update SonataDoctrineORMAdminBundle from 3.5.x to 3.6.0 admin (/admin/app/domainregion/list) dont work. Has an error:
```
Single id is not allowed on composite primary key in entity AppBundle\Entity\DomainOffice
```

## Expected results

I must have the list of entities. It work in 3.5.x version, but i have error in 3.6.0 version SonataDoctrineORMAdminBundle

## Actual results
```
Stack Trace
in vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/MappingException.php at line 424   -
     */
    public static function singleIdNotAllowedOnCompositePrimaryKey($entity)
    {
        return new self('Single id is not allowed on composite primary key in entity '.$entity);
    }
    /**
at MappingException ::singleIdNotAllowedOnCompositePrimaryKey ('AppBundle\Entity\DomainOffice') 
in vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php at line 1780   + 
at ClassMetadataInfo ->getSingleIdentifierFieldName () 
in vendor/doctrine/orm/lib/Doctrine/ORM/Tools/Pagination/CountWalker.php at line 64   + 
at CountWalker ->walkSelectStatement (object(SelectStatement)) 
in vendor/doctrine/orm/lib/Doctrine/ORM/Query/TreeWalkerChain.php at line 113   + 
at TreeWalkerChain ->walkSelectStatement (object(SelectStatement)) 
in vendor/doctrine/orm/lib/Doctrine/ORM/Query/Parser.php at line 380   + 
at Parser ->parse () 
in vendor/doctrine/orm/lib/Doctrine/ORM/Query.php at line 281   + 
at Query ->_parse () 
in vendor/doctrine/orm/lib/Doctrine/ORM/Query.php at line 293   + 
at Query ->_doExecute () 
in vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php at line 962   + 
at AbstractQuery ->executeIgnoreQueryCache (null, '4') 
in vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php at line 917   + 
at AbstractQuery ->execute (null, '4') 
in vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php at line 797   + 
at AbstractQuery ->getSingleResult ('4') 
in vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php at line 826   + 
at AbstractQuery ->getSingleScalarResult () 
in vendor/sonata-project/doctrine-orm-admin-bundle/src/Datagrid/Pager.php at line 48   + 
at Pager ->computeNbResult () 
in vendor/sonata-project/doctrine-orm-admin-bundle/src/Datagrid/Pager.php at line 65   + 
at Pager ->init () 
in vendor/sonata-project/admin-bundle/src/Datagrid/Datagrid.php at line 179   + 
at Datagrid ->buildPager () 
in vendor/sonata-project/admin-bundle/src/Datagrid/Datagrid.php at line 262   + 
at Datagrid ->getForm () 
in vendor/sonata-project/admin-bundle/src/Controller/CRUDController.php at line 167   + 
at CRUDController ->listAction () 
in vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php at line 135   + 
at HttpKernel ->handleRaw (object(Request), '1') 
in vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php at line 57   + 
at HttpKernel ->handle (object(Request), '1', true) 
in vendor/symfony/symfony/src/Symfony/Component/HttpKernel/DependencyInjection/ContainerAwareHttpKernel.php at line 67   + 
at ContainerAwareHttpKernel ->handle (object(Request), '1', true) 
in vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php at line 183   + 
at Kernel ->handle (object(Request)) 
in web/app_dev.php at line 30   + 
```

### PR DESCRIPTION
**Pull request by eugenekurasov**
Added IDENITIY for composite keys for Pager::computeNbResult()

## Subject

Does not work: Admin for entity with a composite keys

Closes #834

## Changelog
```markdown
### Fixed
- Added IDENTITY when CONCAT composite keys
```

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
@eugenekurasov I got your changes but it does not change a thing. :-/

```
Doctrine\ORM\Query\QueryException:
[Semantical Error] line 0, col 61 near 'browserId)))': Error: Invalid PathExpression. Must be a SingleValuedAssociationField.

  at vendor/php/doctrine/orm/lib/Doctrine/ORM/Query/QueryException.php:65
  at Doctrine\ORM\Query\QueryException::semanticalError('line 0, col 61 near \'browserId)))\': Error: Invalid PathExpression. Must be a SingleValuedAssociationField.', object(QueryException))
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/Query/Parser.php:491)
  at Doctrine\ORM\Query\Parser->semanticalError('line 0, col 61 near \'browserId)))\': Error: Invalid PathExpression. Must be a SingleValuedAssociationField.', array('value' => 'browserId', 'type' => 102, 'position' => 61))
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/Query/Parser.php:806)
  at Doctrine\ORM\Query\Parser->processDeferredPathExpressions()
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/Query/Parser.php:271)
  at Doctrine\ORM\Query\Parser->getAST()
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/Query/Parser.php:359)
  at Doctrine\ORM\Query\Parser->parse()
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/Query.php:283)
  at Doctrine\ORM\Query->_parse()
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/Query.php:295)
  at Doctrine\ORM\Query->_doExecute()
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:967)
  at Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache(null, 4)
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:922)
  at Doctrine\ORM\AbstractQuery->execute(null, 4)
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:803)
  at Doctrine\ORM\AbstractQuery->getSingleResult(4)
     (vendor/php/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:831)
  at Doctrine\ORM\AbstractQuery->getSingleScalarResult()
     (vendor/php/sonata-project/doctrine-orm-admin-bundle/src/Datagrid/Pager.php:45)
  at Sonata\DoctrineORMAdminBundle\Datagrid\Pager->computeNbResult()
     (vendor/php/sonata-project/doctrine-orm-admin-bundle/src/Datagrid/Pager.php:62)
  at Sonata\DoctrineORMAdminBundle\Datagrid\Pager->init()
     (vendor/php/sonata-project/admin-bundle/src/Datagrid/Datagrid.php:179)
  at Sonata\AdminBundle\Datagrid\Datagrid->buildPager()
     (vendor/php/sonata-project/admin-bundle/src/Datagrid/Datagrid.php:262)
  at Sonata\AdminBundle\Datagrid\Datagrid->getForm()
     (vendor/php/sonata-project/admin-bundle/src/Controller/CRUDController.php:169)
  at Sonata\AdminBundle\Controller\CRUDController->listAction()
     (vendor/php/symfony/http-kernel/HttpKernel.php:150)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/php/symfony/http-kernel/HttpKernel.php:67)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/php/symfony/http-kernel/Kernel.php:198)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:28)
```

Here is the related entity:

```php
/**
 * UserBrowser
 *
 * @ORM\Table
 * @ORM\Entity
 * @OwnerAware
 */
class UserBrowser implements OwnableInterface
{
    public const COOKIE_NAME = 'browser_id';
    public const UPDATE_GAP = 1800;
    public const NOTIFY_GAP = 30;
    // Used for database clean up
    public const EXPIRATION_GAP = '6 months';
    public const TARGET_SESSION_KEY = 'browser.target_uri';

    /**
     * @var User
     *
     * @ORM\Id
     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\User", inversedBy="browsers")
     * @ORM\JoinColumn(referencedColumnName="id")
     */
    private $user;

    /**
     * @var string
     *
     * @ORM\Id
     * @ORM\Column(type="string", length=64)
     */
    private $browserId;

    /**
     * @var \DateTime
     *
     * @ORM\Column(type="datetime")
     */
    private $usedAt;

    /**
     * @var string|null
     *
     * @ORM\Column(type="string", length=64, nullable=true)
     */
    private $ip;

    /**
     * @var string|null
     *
     * @ORM\Column(type="string", length=255, nullable=true)
     */
    private $reverse;

    /**
     * @var string|null
     *
     * @ORM\Column(type="string", length=64, nullable=true)
     */
    private $name;

    /**
     * @var string|null
     *
     * @ORM\Column(type="string", length=6, nullable=true)
     */
    private $verificationCode;

    /**
     * @var \DateTime|null
     *
     * @ORM\Column(type="datetime", nullable=true)
     */
    private $verificationAskedAt;

    /**
     * @var mixed[]
     *
     * @ORM\Column(type="array")
     */
    private $browserDetails = [];
```

**Comment by Soullivaneuh** [comment]
Here is the generated DQL:

```
SELECT count(DISTINCT concat(IDENTITY(o.user),'-',IDENTITY(o.browserId))) as cnt FROM AppBundle\Entity\UserBrowser o
```

**Comment by eugenekurasov** [comment]
@Soullivaneuh, thanks for feedback. Problem is difficult than I thought. I will continue fix.

**Comment by Soullivaneuh** [comment]
Hi @eugenekurasov,

Do you have anything new? 

If no, I'd like to revert the original feature and re-open the related issue.

It's more important to not break things than add features, and you will have plenty of time to re-done your PR considering composite keys with tests, if you want.

**Comment by eugenekurasov** [comment]
@Soullivaneuh  hi, I will be continue on weekend.

I can create revert commit. On this [PR](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/844) has been revert **and fixed wrong count** of composite keys (https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/844#issuecomment-415853484)

I can revert with wrong count and this will fix exception. And continue fix wrong count on composite keys.

**Comment by Soullivaneuh** [comment]
@eugenekurasov I have to confess I didn't followed all the process.

I only know it was not broke before v3.6.0.

So yeah, if you think you need more time, I would suggest a revert of what is causing trouble and then create an another PR to re-introduce the feature with a fresh start. :+1: 

Are you able at least to create a proof test with the information I sent to you?

**Comment by eugenekurasov** [comment]
I have did mistake with union two issues in one PR(https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/827):
- Fix DISTINCT by slow query
- Fix wrong count on composite keys.

**Comment by eugenekurasov** [comment]
@Soullivaneuh yes, your information help me. But me need more time for fix. I can create revert commit with tests.

File: tests/Datagrid/PagerTest.php [hunk 1]
```diff
@@ -14,13 +14,39 @@
...
+        $schemaTool->createSchema($classes);
+
+        $qb = $em->createQueryBuilder()
+            ->select('o')
+            ->from(StoreProduct::class, 'o');
+        $pq = new ProxyQuery($qb);
+        $pager = new Pager();
+        $pager->setCountColumn($em->getClassMetadata(StoreProduct::class)->getIdentifierFieldNames());
+        $pager->setQuery($pq);
+        $this->assertEquals(0, $pager->computeNbResult());
```
Comment (phansys) [inline tests/Datagrid/PagerTest.php]:
```suggestion
        $this->assertSame(0, $pager->computeNbResult());
```

File: tests/Datagrid/PagerTest.php [hunk 2]
```diff
@@ -14,13 +14,39 @@
...
 use Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQuery;
 use Sonata\DoctrineORMAdminBundle\Tests\Filter\QueryBuilder;
+use Sonata\DoctrineORMAdminBundle\Tests\Fixtures\Entity\ORM\Product;
+use Sonata\DoctrineORMAdminBundle\Tests\Fixtures\Entity\ORM\Store;
+use Sonata\DoctrineORMAdminBundle\Tests\Fixtures\Entity\ORM\StoreProduct;
+use Symfony\Bridge\Doctrine\Test\DoctrineTestHelper;
 
 class PagerTest extends TestCase
 {
+    public function testComputeNbResultFoCompositeId()
```
Comment (phansys) [inline tests/Datagrid/PagerTest.php]:
```suggestion
    public function testComputeNbResultFoCompositeId(): void
```

**Comment by phansys** [comment]
Regarding the issues around composite keys, #928 was merged recently. I don't know if this PR is related, but we should take that into account.

**Pull request closed without merging**